File: app/src/test/java/dev/suhockii/lifetest/repo/LocalRepositoryTest.java
package dev.suhockii.lifetest.repo;

import com.facebook.stetho.inspector.elements.Document;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.Mockito;

import java.util.ArrayList;
import java.util.List;

import dev.suhockii.lifetest.data.local.dao.ProductDao;
import dev.suhockii.lifetest.data.local.dao.ProductDetailsDao;
import dev.suhockii.lifetest.data.local.entity.ProductDetailsEntity;
import dev.suhockii.lifetest.data.local.entity.ProductEntity;
import dev.suhockii.lifetest.data.remote.entity.ProductDetailsResponse;
import dev.suhockii.lifetest.model.Product;
import dev.suhockii.lifetest.model.ProductDetails;
import io.reactivex.Observable;
import io.reactivex.Single;

import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(JUnit4.class)
public class LocalRepositoryTest {
    private ProductDao productDao;
    private ProductDetailsDao productDetailsDao;
    private AppRepository localRepository;

    @Before
    public void setUp() throws Exception {
        productDao = Mockito.mock(ProductDao.class);
        productDetailsDao = Mockito.mock(ProductDetailsDao.class);
        localRepository = new LocalRepository(productDao, productDetailsDao);
    }

    @Test
    public void getProducts() throws Exception {
        when(productDao.getProducts()).thenReturn(Single.just(new ArrayList<>()));

        localRepository.getProducts();

        verify(productDao, times(1)).getProducts();
    }

    @Test
    public void getDetailsFor() throws Exception {
        String productId = "1";
        ProductDetailsEntity productDetails = new ProductDetailsEntity(productId);
        Product product = new ProductEntity(productId);

        when(productDetailsDao.getProductDetails(productId)).thenReturn(Single.just(productDetails));

        localRepository.getDetailsFor(product);

        verify(productDetailsDao, times(1)).getProductDetails(productId);
    }

    @SuppressWarnings("unchecked")
    @Test
    public void saveProducts() throws Exception {
        ArrayList<Product> products = new ArrayList<>();
        List<ProductEntity> productEntities = (List<ProductEntity>) (List<?>) products;

        localRepository.saveProducts(products);

        verify(productDao, times(1)).saveProducts(productEntities);
    }

    @Test
    public void saveProductDetails() throws Exception {
        String productId = "1";
        ProductDetails productDetails = new ProductDetailsResponse(productId);
        ProductDetailsEntity productDetailsEntity = new ProductDetailsEntity.Builder()
                .id(productDetails.getId())
                .build();

        localRepository.saveProductDetails(productDetails);

        verify(productDetailsDao, times(1)).saveProductDetails(productDetailsEntity);
    }

}

File: app/src/test/java/dev/suhockii/lifetest/repo/LocalRepositoryTest.kt
package dev.suhockii.lifetest.repo

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.Mockito

import java.util.ArrayList

import dev.suhockii.lifetest.data.local.dao.ProductDao
import dev.suhockii.lifetest.data.local.dao.ProductDetailsDao
import dev.suhockii.lifetest.data.local.entity.ProductDetailsEntity
import dev.suhockii.lifetest.data.local.entity.ProductEntity
import dev.suhockii.lifetest.data.remote.entity.ProductDetailsResponse
import dev.suhockii.lifetest.model.Product
import dev.suhockii.lifetest.model.ProductDetails
import io.reactivex.Single

import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`

@RunWith(JUnit4::class)
class LocalRepositoryTest {
    private var productDao: ProductDao? = null
    private var productDetailsDao: ProductDetailsDao? = null
    private var localRepository: AppRepository? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        productDao = Mockito.mock(ProductDao::class.java)
        productDetailsDao = Mockito.mock(ProductDetailsDao::class.java)
        localRepository = LocalRepository(productDao!!, productDetailsDao!!)
    }

    @Test
    @Throws(Exception::class)
    fun getProducts() {
        `when`(productDao!!.products).thenReturn(Single.just(ArrayList()))

        localRepository!!.getProducts()

        verify<ProductDao>(productDao, times(1)).products
    }

    @Test
    @Throws(Exception::class)
    fun getDetailsFor() {
        val productId = "1"
        val productDetails = ProductDetailsEntity(productId)
        val product = ProductEntity(productId)

        `when`(productDetailsDao!!.getProductDetails(productId)).thenReturn(Single.just(productDetails))

        localRepository!!.getDetailsFor(product)

        verify<ProductDetailsDao>(productDetailsDao, times(1)).getProductDetails(productId)
    }

    @Test
    @Throws(Exception::class)
    fun saveProducts() {
        val products = ArrayList<Product>()
        val productEntities = products as List<ProductEntity>

        localRepository!!.saveProducts(products)

        verify<ProductDao>(productDao, times(1)).saveProducts(productEntities)
    }

    @Test
    @Throws(Exception::class)
    fun saveProductDetails() {
        val productId = "1"
        val productDetails = ProductDetailsResponse(productId)
        val productDetailsEntity = ProductDetailsEntity.Builder()
            .id(productDetails.id)
            .build()

        localRepository!!.saveProductDetails(productDetails)

        verify<ProductDetailsDao>(productDetailsDao, times(1)).saveProductDetails(productDetailsEntity)
    }

}